processor/signoztailsampler/evaluator_test.go

```diff
@@ -0,0 +1,341 @@
+		// store initiat decision for each policy
+		initialDecisions := make([]sampling.Decision, 1)
+		// there is only one policy, so set initiatDecisions
+		// for this policy to pending
+		initialDecisions[0] = sampling.Pending
+
+		traceData := &sampling.TraceData{
+			Decisions:       initialDecisions,
+			ArrivalTime:     time.Now(),
+			SpanCount:       atomic.NewInt64(1),
+			ReceivedBatches: td,
+		}
+
+		return e.Evaluate(id, traceData)
+	}
+
+	decision, err := eval(traceIds[0], traces[0])
+
+	require.NoError(t, err, "failed to evalute trace for policy")
+	require.Equal(t, sampling.Sampled, decision, "expected to sample a valid trace as per policy")
+
+	traceIds1, traces1 := generateTraceForTestEval(1, "source", "none")
+	decision1, err1 := eval(traceIds1[0], traces1[0])
+
+	require.NoError(t, err1, "failed to evalute trace for policy")
+	require.Equal(t, sampling.NoResult, decision1, "expected to not sampled decision for trace as per policy")
+}
+
+func TestProbablisticCfg(t *testing.T) {
+	cfg := PolicyCfg{
+		Name: "security",
+		ProbabilisticCfg: ProbabilisticCfg{
+			SamplingPercentage: 100,
+		},
+		PolicyFilterCfg: PolicyFilterCfg{
+			StringAttributeCfgs: []StringAttributeCfg{
+				{
+					Key:    "source",
+					Values: []string{"audit"},
+				},
+			},
+		},
+	}
+
+	e := NewDefaultEvaluator(zap.NewNop(), &cfg)
+	traceIds, traces := generateTraceForTestEval(1, "source", "audit")
+
+	eval := func(e sampling.PolicyEvaluator, id pcommon.TraceID, td ptrace.Traces) (sampling.Decision, error) {
+		// store initiat decision for each policy
+		initialDecisions := make([]sampling.Decision, 1)
+		// there is only one policy, so set initiatDecisions
+		// for this policy to pending
+		initialDecisions[0] = sampling.Pending
+
+		traceData := &sampling.TraceData{
+			Decisions:       initialDecisions,
+			ArrivalTime:     time.Now(),
+			SpanCount:       atomic.NewInt64(1),
+			ReceivedBatches: td,
+		}
+
+		return e.Evaluate(id, traceData)
+	}
+
+	decision, err := eval(e, traceIds[0], traces[0])
+
+	require.NoError(t, err, "failed to evalute trace for policy")
+	require.Equal(t, sampling.Sampled, decision, "expected to sample a valid trace as per policy")
+
+	cfg.SamplingPercentage = 0
+	zeroSamplingEvaluator := NewDefaultEvaluator(zap.NewNop(), &cfg)
+	decision1, err1 := eval(zeroSamplingEvaluator, traceIds[0], traces[0])
+
+	require.NoError(t, err1, "failed to evaluate trace for policy with zero sampling")
+	require.Equal(t, sampling.NotSampled, decision1, "expected to not sampled as sampling percent is zero")
+
+	cfg.SamplingPercentage = 99
+	evaluator99 := NewDefaultEvaluator(zap.NewNop(), &cfg)
+	decision99, err99 := eval(evaluator99, traceIds[0], traces[0])
+
+	require.NoError(t, err99, "failed to evaluate trace for policy with 99 percent sampling")
+	require.Equal(t, sampling.Sampled, decision99, "expected to sampled as sampling percent is 99, re-run the test to try again")
```

Comment:
Please max of X retries here itself to reduce the flakiness of tests.

Reply:
I am not sure how to handle this. Please do suggest the approach in detail. 
I can increase the sampling rate to 99.99 or re-try a few number of times to see majority of them succeed.